File: backend/src/main/java/jp/tubeboard/features/lives/pdf/canvas/CanvasSchema.java
package jp.tubeboard.features.lives.pdf.canvas;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import jp.tubeboard.features.lives.pdf.PdfOrientation;
import jp.tubeboard.features.lives.pdf.PdfPaperSize;

/**
 * Schema for the canvas-based PDF layout. Coordinates are millimeters with the
 * origin at the top-left of the page (PowerPoint convention).
 */
public final class CanvasSchema {

    private CanvasSchema() {
    }

    public record CanvasDocument(CanvasPage page, List<CanvasElement> elements) {
    }

    public record CanvasPage(
            PdfPaperSize size,
            PdfOrientation orientation,
            Float marginMm,
            Float baseFontSizePt) {
    }

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "kind")
    @JsonSubTypes({
            @JsonSubTypes.Type(value = CanvasElement.TextElement.class, name = "text"),
            @JsonSubTypes.Type(value = CanvasElement.FieldElement.class, name = "field"),
            @JsonSubTypes.Type(value = CanvasElement.DividerElement.class, name = "divider"),
            @JsonSubTypes.Type(value = CanvasElement.SpacerElement.class, name = "spacer"),
            @JsonSubTypes.Type(value = CanvasElement.TableElement.class, name = "table"),
    })
    public sealed interface CanvasElement {

        String id();

        Float xMm();

        Float yMm();

        Float wMm();

        Float hMm();

        record TextElement(String id, Float xMm, Float yMm, Float wMm, Float hMm,
                String content, Float fontSizePt, Boolean bold, Boolean italic,
                String align, String verticalAlign, String color,
                String backgroundColor, String borderColor, Float borderThicknessPt)
                implements CanvasElement {
        }

        record FieldElement(String id, Float xMm, Float yMm, Float wMm, Float hMm,
                String fieldId, String fallbackLabel, Boolean showLabel, String labelOverride,
                Float fontSizePt, Boolean bold, String align, String verticalAlign, String color,
                String backgroundColor, String borderColor, Float borderThicknessPt)
                implements CanvasElement {
        }

        record DividerElement(String id, Float xMm, Float yMm, Float wMm, Float hMm,
                String color, Float thicknessPt)
                implements CanvasElement {
        }

        record SpacerElement(String id, Float xMm, Float yMm, Float wMm, Float hMm)
                implements CanvasElement {
        }

        /**
         * @param autoGrow when not {@code false} the table grows downwards to fit
         *                 its rows, pushing later elements down and continuing on
         *                 the next page when it no longer fits. {@code hMm} then
         *                 acts as a minimum height.
         */
        record TableElement(String id, Float xMm, Float yMm, Float wMm, Float hMm,
                TableSource source, List<TableColumn> columns, Boolean showHeader,
                Float fontSizePt, String headerFill, String borderColor, Boolean zebra,
                Boolean autoGrow)
                implements CanvasElement {
        }
    }

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "kind")
    @JsonSubTypes({
            @JsonSubTypes.Type(value = TableSource.GroupSource.class, name = "group"),
            @JsonSubTypes.Type(value = TableSource.FieldsSource.class, name = "fields"),
    })
    public sealed interface TableSource {

        record GroupSource(String groupId, String fallbackLabel) implements TableSource {
        }

        record FieldsSource(List<FieldRef> fields) implements TableSource {
        }

        record FieldRef(String fieldId, String fallbackLabel) {
        }
    }

    /**
     * @param shrinkToFit    when {@code true} a cell whose text is too wide for
     *                       the column is drawn at a reduced font size instead of
     *                       being wrapped. Text that still does not fit at
     *                       {@code minFontSizePt} wraps as before.
     * @param minFontSizePt  lower bound for the shrinking above (defaults to 6pt).
     */
    public record TableColumn(
            String id,
            String header,
            String fieldId,
            Float widthRatio,
            String align,
            String format,
            Boolean shrinkToFit,
            Float minFontSizePt) {
    }
}

File: frontend/src/features/lives/pdf/canvas-schema.ts
/** TypeScript twin of the backend `CanvasSchema` records. Keep in sync. */

export type PaperSize = 'A3' | 'A4' | 'A5' | 'B4' | 'B5' | 'LETTER';
export type Orientation = 'PORTRAIT' | 'LANDSCAPE';
export type HAlign = 'left' | 'center' | 'right';
export type VAlign = 'top' | 'middle' | 'bottom';

export interface CanvasPage {
  size: PaperSize;
  orientation: Orientation;
  marginMm: number;
  baseFontSizePt: number;
}

interface BaseElement {
  id: string;
  xMm: number;
  yMm: number;
  wMm: number;
  hMm: number;
}

export interface TextElement extends BaseElement {
  kind: 'text';
  content: string;
  fontSizePt: number;
  bold?: boolean;
  italic?: boolean;
  align?: HAlign;
  verticalAlign?: VAlign;
  color?: string;
  backgroundColor?: string;
  borderColor?: string;
  borderThicknessPt?: number;
}

export interface FieldElement extends BaseElement {
  kind: 'field';
  fieldId: string;
  fallbackLabel?: string;
  showLabel?: boolean;
  labelOverride?: string;
  fontSizePt: number;
  bold?: boolean;
  align?: HAlign;
  verticalAlign?: VAlign;
  color?: string;
  backgroundColor?: string;
  borderColor?: string;
  borderThicknessPt?: number;
}

export interface DividerElement extends BaseElement {
  kind: 'divider';
  color?: string;
  thicknessPt?: number;
}

export interface SpacerElement extends BaseElement {
  kind: 'spacer';
}

export type TableSource =
  | { kind: 'group'; groupId: string; fallbackLabel?: string }
  | { kind: 'fields'; fields: Array<{ fieldId: string; fallbackLabel?: string }> };

export interface TableColumn {
  id: string;
  header: string;
  /** Field id from the form, or `__index__` (group rows) / `__label__` (fields rows). */
  fieldId: string;
  widthRatio: number;
  align?: HAlign;
  format?: string;
  /** 列幅に収まらないセルを折り返す代わりに文字を縮める。 */
  shrinkToFit?: boolean;
  /** 上の縮小の下限。既定 6pt。ここまで縮めても収まらない分は折り返す。 */
  minFontSizePt?: number;
}

export interface TableElement extends BaseElement {
  kind: 'table';
  source: TableSource;
  columns: TableColumn[];
  showHeader?: boolean;
  fontSizePt: number;
  headerFill?: string;
  borderColor?: string;
  zebra?: boolean;
  /**
   * 既定 true。行が入り切らないときに下方向へ伸び、下の要素を押し下げ、
   * ページに収まらなくなったら次ページに続きを出す。このとき hMm は最低高さになる。
   */
  autoGrow?: boolean;
}

export const DEFAULT_MIN_FONT_SIZE_PT = 6;

export type CanvasElement = TextElement | FieldElement | DividerElement | SpacerElement | TableElement;

export interface CanvasDocument {
  page: CanvasPage;
  elements: CanvasElement[];
}

export const PAPER_SIZE_OPTIONS: Array<{ value: PaperSize; label: string; widthMm: number; heightMm: number }> = [
  { value: 'A3', label: 'A3', widthMm: 297, heightMm: 420 },
  { value: 'A4', label: 'A4', widthMm: 210, heightMm: 297 },
  { value: 'A5', label: 'A5', widthMm: 148, heightMm: 210 },
  { value: 'B4', label: 'B4 (JIS)', widthMm: 257, heightMm: 364 },
  { value: 'B5', label: 'B5 (JIS)', widthMm: 182, heightMm: 257 },
  { value: 'LETTER', label: 'Letter', widthMm: 216, heightMm: 279 },
];

export const ORIENTATION_OPTIONS: Array<{ value: Orientation; label: string }> = [
  { value: 'LANDSCAPE', label: '横' },
  { value: 'PORTRAIT', label: '縦' },
];

export function getPaperDimensions(size: PaperSize, orientation: Orientation): { widthMm: number; heightMm: number } {
  const opt = PAPER_SIZE_OPTIONS.find((o) => o.value === size) ?? PAPER_SIZE_OPTIONS[1];
  return orientation === 'LANDSCAPE'
    ? { widthMm: opt.heightMm, heightMm: opt.widthMm }
    : { widthMm: opt.widthMm, heightMm: opt.heightMm };
}

export function elementKindLabel(kind: CanvasElement['kind']): string {
  switch (kind) {
    case 'text': return 'テキスト';
    case 'field': return 'フィールド';
    case 'divider': return '区切り線';
    case 'spacer': return 'スペーサー';
    case 'table': return '表';
  }
}
